src/expr/src/expr/expr_array_concat.rs

```diff
@@ -0,0 +1,502 @@
+            .finish()
+    }
+}
+
+impl ArrayConcatExpression {
+    pub fn new(
+        return_type: DataType,
+        left: BoxedExpression,
+        right: BoxedExpression,
+        op: Operation,
+    ) -> Self {
+        Self {
+            return_type,
+            left,
+            right,
+            op,
+            op_func: match op {
+                Operation::ConcatArray => Self::concat_array,
+                Operation::AppendArray => Self::append_array,
+                Operation::PrependArray => Self::prepend_array,
+                Operation::AppendValue => Self::append_value,
+                Operation::PrependValue => Self::prepend_value,
+            },
+        }
+    }
+
+    /// Concatenates two arrays with same data type.
+    /// The behavior is the same as PG.
+    ///
+    /// Examples:
+    /// - `select array_cat(array[66], array[123]);` => `[66,123]`
+    /// - `select array_cat(array[66], null::int[]);` => `[66]`
+    /// - `select array_cat(null::int[], array[123]);` => `[123]`
+    fn concat_array(left: DatumRef, right: DatumRef) -> Datum {
+        match (left, right) {
+            (None, right) => right.map(ScalarRefImpl::into_scalar_impl),
+            (left, None) => left.map(ScalarRefImpl::into_scalar_impl),
+            (Some(ScalarRefImpl::List(left)), Some(ScalarRefImpl::List(right))) => Some(
+                ListValue::new(
+                    left.values_ref()
+                        .into_iter()
+                        .chain(right.values_ref().into_iter())
+                        .map(|x| x.map(ScalarRefImpl::into_scalar_impl))
+                        .collect(),
+                )
+                .into(),
+            ),
+            _ => {
+                panic!("the operands must be two arrays with the same data type");
+            }
+        }
+    }
+
+    /// Appends an array as the back element of an array of array.
+    /// Note the behavior is slightly different from PG.
+    ///
+    /// Examples:
+    /// - `select array_cat(array[array[66]], array[233]);` => `[[66], [233]]`
+    /// - `select array_cat(array[array[66]], null::int[]);` => `[[66]]` ignore NULL, same as PG
+    /// - `select array_cat(null::int[][], array[233]);` => `[[233]]` different from PG
+    /// - `select array_cat(null::int[][], null::int[]);` => `null` same as PG
+    fn append_array(left: DatumRef, right: DatumRef) -> Datum {
+        match (left, right) {
+            (None, None) => None,
+            (None, right) => {
+                Some(ListValue::new(vec![right.map(ScalarRefImpl::into_scalar_impl)]).into())
+            }
+            (left, None) => left.map(ScalarRefImpl::into_scalar_impl),
+            (Some(ScalarRefImpl::List(left)), right) => Some(
+                ListValue::new(
+                    left.values_ref()
+                        .into_iter()
+                        .chain(std::iter::once(right))
+                        .map(|x| x.map(ScalarRefImpl::into_scalar_impl))
```

Comment:
If we support storing an iterator of `DatumRef` in the `ListRef`, then we can use `builder.append_datum_ref` to acheive zero-copy. 🤣 But it may require some hacks.